dev/iterator.h

```diff
@@ -1,101 +1,120 @@
+                if(!this->current) {
+                    this->stmt.reset();
+                }
             }
 
             void next() {
                 this->current.reset();
-                if(sqlite3_stmt* stmt = this->stmt.get()) {
-                    perform_step(stmt, std::bind(&iterator_t::extract_value, this));
-                    if(!this->current) {
-                        this->stmt.reset();
-                    }
-                }
+                this->step();
             }
 
           public:
-            using difference_type = ptrdiff_t;
-            using pointer = value_type*;
-            using reference = value_type&;
-            using iterator_category = std::input_iterator_tag;
+            iterator_t() = default;
 
-            iterator_t(){};
-
-            iterator_t(statement_finalizer stmt_, view_type& view_) : stmt{std::move(stmt_)}, view{&view_} {
-                next();
+            iterator_t(const db_objects_type& dbObjects, statement_finalizer stmt) :
+                db_objects{&dbObjects}, stmt{std::move(stmt)} {
+                this->step();
             }
 
+            iterator_t(const iterator_t&) = default;
+            iterator_t& operator=(const iterator_t&) = default;
+            iterator_t(iterator_t&&) = default;
+            iterator_t& operator=(iterator_t&&) = default;
+
             value_type& operator*() const {
-                if(!this->stmt || !this->current) {
-                    throw std::system_error{orm_error_code::trying_to_dereference_null_iterator};
-                }
+                if(!this->stmt)
+                    SQLITE_ORM_CPP_UNLIKELY {
+                        throw std::system_error{orm_error_code::trying_to_dereference_null_iterator};
+                    }
                 return *this->current;
             }
 
+            // note: should actually be only present for contiguous iterators
             value_type* operator->() const {
                 return &(this->operator*());
             }
 
-            iterator_t<V>& operator++() {
+            iterator_t& operator++() {
                 next();
                 return *this;
             }
 
-            void operator++(int) {
-                this->operator++();
+            iterator_t operator++(int) {
+                auto tmp = *this;
+                ++*this;
+                return tmp;
             }
 
-            bool operator==(const iterator_t& other) const {
-                return this->current == other.current;
+            friend bool operator==(const iterator_t& lhs, const iterator_t& rhs) {
+                return lhs.current == rhs.current;
             }
 
-            bool operator!=(const iterator_t& other) const {
-                return !(*this == other);
+#ifndef SQLITE_ORM_DEFAULT_COMPARISONS_SUPPORTED
+            friend bool operator!=(const iterator_t& lhs, const iterator_t& rhs) {
```

Comment:
Are you sure we need `friend` here as well? We don't access `current` field here so we can skip `friend`

Reply:
Yes, we do need `friend` as it is a global operator...